src/sources/util/http.rs

```diff
@@ -0,0 +1,110 @@
+use crate::event::Event;
+use bytes::Buf;
+use futures::{sync::mpsc, Future, Sink};
+use serde::Serialize;
+use std::error::Error;
+use std::fmt::{self, Display};
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex};
+use stream_cancel::Tripwire;
+use warp::filters::BoxedFilter;
+use warp::http::{HeaderMap, StatusCode};
+use warp::{Filter, Rejection};
+
+#[derive(Serialize, Debug)]
+pub struct ErrorMessage {
+    code: u16,
+    message: String,
+}
+impl ErrorMessage {
+    pub fn new(code: StatusCode, message: String) -> Self {
+        ErrorMessage {
+            code: code.as_u16(),
+            message,
+        }
+    }
+}
+impl Error for ErrorMessage {}
+impl Display for ErrorMessage {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}: {}", self.code, self.message)
+    }
+}
+
+fn string_to_static_str(s: String) -> &'static str {
+    //necessary because warp 0.1.18 needs a &'static str for the path
+    Box::leak(s.into_boxed_str())
+}
+
+pub trait HttpSource: Clone + Send + Sync + 'static {
+    fn build_event(
+        &self,
+        body: impl Buf,
+        header_map: HeaderMap,
+    ) -> Result<Vec<Event>, ErrorMessage>;
+
+    fn run(
+        self,
+        address: SocketAddr,
+        path: String,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<crate::sources::Source> {
+        let (trigger, tripwire) = Tripwire::new();
+        let trigger = Arc::new(Mutex::new(Some(trigger)));
+
+        let mut filter: BoxedFilter<()> = warp::post2().boxed();
+        if !path.is_empty() && path != "/" {
+            for s in string_to_static_str(path).split('/') {
+                filter = filter.and(warp::path(s)).boxed();
+            }
+        }
+        let svc = filter
+            .and(warp::path::end())
+            .boxed()
+            .and(warp::header::headers_cloned())
+            .and(warp::body::concat())
+            .and_then(move |headers: HeaderMap, body| {
+                let out = out.clone();
+                let trigger = trigger.clone();
+                info!("Handling http request: {:?}", headers);
+
+                futures::future::result(
+                    self.build_event(body, headers)
+                        .map_err(warp::reject::custom),
+                )
+                .and_then(|events| {
+                    out.send_all(futures::stream::iter_ok(events)).map_err(
+                        move |_: mpsc::SendError<Event>| {
```

Comment:
Is there any context we can get from this error? It'd be sad to throw it away.

Reply:
We sort of can get more context.
We can log the event it was trying to send when it failed (which is a good idea), but the only reason it would fail is if the receiver is dropped, and the error doesn't give any more insight than that.

Reply:
OK, fair!